docs/design_doc.md

```diff
@@ -0,0 +1,75 @@
+
+The current design incurs a significant overhead during redeployment as each replica has to build and maintain its own copy of the cache. This design does not scale well if the number of replicas is increased, as the [hourly quota](https://developer.github.com/v3/#rate-limiting) for Github API requests will quickly be exhausted, causing a hold up for all [clients](#client) depending on this service.
+
+## Proposed solution
+
+* Replace each pod's in-memory cache with a single shared cache for all pods. 
+
+    * The [cache server](#cache-server) should be deployed within the same Open Shift cluster as the Github mirror to reduce the latency associated with requests to the cache.
+
+    * A shared cache should reduce the number of calls to the Github API since each pod will not maintain its own copy of the cache.
+
+    * When the mirror is redeployed, the cache will have to be rebuilt once only, instead of once for *each* replica.
+
+* A highly available key-value store such as Redis can be used as the cache server.
+
+    * The [etags](https://developer.github.com/v3/#conditional-requests) returned with each conditional request to the Github API are unique 
+
+    * Redis does not guarantee [strong consistency](https://redis.io/topics/cluster-tutorial#redis-cluster-consistency-guarantees) which is acceptable since any stale cache reads will always be compared against the response from the conditional request, resulting - in the worst case - in an additional request to the Github api as well as unnecessairly replacing the cache entry.
+
+This redesign should enable multi-processing for the mirror since different processes can access the shared cache server which is currently not possible when using an in-memory data structure.
+
+### Architecture diagram
+
+![](./images/arch.png)
+
+## Scope
+
+* Backing up the cache is not necessary as long as the cache rebuild time is constant in the size of mirror replicas.
+
+* Project should use free and open source software and not incur an additional dollar cost.
+
+## Acceptance Criteria
+
+* A successful solution will provide a highly available, low latency mirror server, that can be easily scaled from 3 to 100 replicas.
+
+* Additional latency due to redesign will be < 10 ms and total request time will not exceed 3 [http request times](http://services.google.com/fh/files/blogs/google_delayexp.pdf).
+
+* User can set up the mirror to use either the per-pod in-memory cache or the Redis server simply by modifying a configuration file. 
+
+## Extensions
+
+* Create a backup of the cache to avoid unecessairly rebuilding it in the case or a server/pod failure or redeployment. One solution would involve rebuilding the cache from snapshots stored in an external database.
+
+* To increase availability and to keep the cache from growing too large, sharding can be used to break up the cache into smaller parts stored on separate shards.
```

Comment:
We should also set some expiration time for the cache objects, so we don't keep forever resources that are no longer used. Our integrations, the main users of the github mirror, will execute 6 times per hour in average. If a resource is not read in say 1h, it would be safe to drop it I guess (I'm not saying 1h is a good value for the expiration time - we can be very conservative here).

Reply:
added.